File: cases/taint/javascript/function-field-positive/FunctionFieldPositive.js
function dfb_source() { // DFB-SOURCE: function-field-input
  return "tainted";
}

function dfb_sink(value) {} // DFB-SINK: function-field-sink

function invoke(target, value) {
  target.fn(value);
}

function run() {
  const holder = { fn: null };
  const otherHolder = { fn: null };
  holder.fn = (value) => { // DFB-WITNESS: function-field-store
    dfb_sink(value);
  };
  otherHolder.fn = (value) => {
    dfb_sink("clean");
  };
  invoke(holder, dfb_source());
}

File: cases/taint/cpp/function-field-positive/function_field_positive.cpp
#include <functional>
#include <string>

std::string dfb_source() { // DFB-SOURCE: function-field-input
    return "tainted";
}

void dfb_sink(const std::string &value) {} // DFB-SINK: function-field-sink

struct Holder {
    std::function<void(const std::string &)> fn;
};

void invoke(const Holder &holder, const std::string &value) {
    holder.fn(value);
}

void run() {
    Holder leaking;
    leaking.fn = [](const std::string &value) { dfb_sink(value); }; // DFB-WITNESS: function-field-store
    Holder dropping;
    dropping.fn = [](const std::string &value) { dfb_sink("clean"); };
    invoke(leaking, dfb_source());
}
